client/starwhale/api/_impl/dataset/loader.py

```diff
@@ -172,20 +173,81 @@ def __init__(
+        self._meta_fetched_queue: queue.Queue[_TMetaQItem] | None = None
+        self._row_unpacked_queue: queue.Queue[_TRowQItem] | None = None
+        self._key_processed_queue: queue.Queue[_TProcessedQItem] | None = None
+
+        self._key_range_dict: t.Dict[t.Tuple[t.Any, t.Any], t.Dict[str, int]] = {}
+
+    def _check_pipe_data_empty(self) -> bool:
+        if (
+            self._meta_fetched_queue is not None
+            and self._meta_fetched_queue.qsize() > 0
+        ):
+            return False
+
+        if (
+            self._row_unpacked_queue is not None
+            and self._row_unpacked_queue.qsize() > 0
+        ):
+            return False
+
+        if (
+            self._key_processed_queue is not None
+            and self._key_processed_queue.qsize() > 0
+        ):
+            return False
+
+        return True
+
+    def _get_processed_key_range(self) -> t.Optional[t.List[t.Tuple[t.Any, t.Any]]]:
+        if self._key_processed_queue is None:
+            raise RuntimeError("key processed queue is not initialized")
+
+        # Current server side implementation only supports the original key range as the processedData parameter,
+        # so we need to wait for all the keys in the original key range to be processed.
+        while not self._key_processed_queue.empty():
+            key = self._key_processed_queue.get(block=True)
+
+            # TODO: tune performance for find key range
+            for rk in self._key_range_dict:
+                if (rk[0] is None or rk[0] <= key) and (rk[1] is None or key < rk[1]):
+                    self._key_range_dict[rk]["processed_cnt"] += 1
+                    break
+            else:
+                raise RuntimeError(
+                    f"key({key}) not found in key range dict:{self._key_range_dict}"
+                )
+
+        processed_range_keys = []
+        for rk in list(self._key_range_dict.keys()):
+            if (
+                self._key_range_dict[rk]["processed_cnt"]
+                == self._key_range_dict[rk]["rows_cnt"]
+            ):
+                processed_range_keys.append(rk)
+                del self._key_range_dict[rk]
+
+        return processed_range_keys
+
     def _iter_meta(self) -> t.Generator[TabularDatasetRow, None, None]:
         if not self.session_consumption:
             # TODO: refactor for batch-signed urls
             for row in self.tabular_dataset.scan():
                 yield row
         else:
             while True:
-                # TODO: tune last processed range for multithread
-                pk = [self.last_processed_range] if self.last_processed_range else None
+                pk = self._get_processed_key_range()
                 rt = self.session_consumption.get_scan_range(pk)
-                self.last_processed_range = rt
                 if rt is None:
-                    break
+                    if self._iter_row_processing or not self._check_pipe_data_empty():
```

Comment:
Add a lock to ensure that the queues will not be updated during the check time?

Reply:
updated